packages/react-dom-bindings/src/client/ReactDOMFloatClient.js

```diff
@@ -0,0 +1,825 @@
+  ) {
+    const as = options.as;
+    const resource = preloadResources.get(href);
+    if (resource) {
+      if (__DEV__) {
+        const originallyImplicit =
+          (resource: any)._dev_implicit_construction === true;
+        const latestProps = preloadPropsFromPreloadOptions(href, as, options);
+        validatePreloadResourceDifference(
+          resource.props,
+          originallyImplicit,
+          latestProps,
+          false,
+        );
+      }
+    } else {
+      const resourceProps = preloadPropsFromPreloadOptions(href, as, options);
+      createPreloadResource(ownerDocument, href, resourceProps);
+    }
+  }
+}
+
+function preloadPropsFromPreloadOptions(
+  href: string,
+  as: ResourceType,
+  options: PreloadOptions,
+): PreloadProps {
+  return {
+    href,
+    rel: 'preload',
+    as,
+    crossOrigin: as === 'font' ? '' : options.crossOrigin,
+  };
+}
+
+// --------------------------------------
+//      ReactDOM.preinit
+// --------------------------------------
+
+type PreinitAs = 'style';
+type PreinitOptions = {
+  as: PreinitAs,
+  crossOrigin?: string,
+  precedence?: string,
+};
+function preinit(href: string, options: PreinitOptions) {
+  if (__DEV__) {
+    validatePreinitArguments(href, options);
+  }
+
+  if (
+    typeof href === 'string' &&
+    href &&
+    typeof options === 'object' &&
+    options !== null
+  ) {
+    const as = options.as;
+    if (!currentDocument) {
+      // We are going to emit a preload as a best effort fallback since this preinit
+      // was called outside of a render. Given the passive nature of this fallback
+      // we do not warn in dev when props disagree if there happens to already be a
+      // matching preload with this href
+      const preloadDocument = getDocumentForPreloads();
+      if (preloadDocument) {
+        const preloadResource = preloadResources.get(href);
+        if (!preloadResource) {
+          const preloadProps = preloadPropsFromPreinitOptions(
+            href,
+            as,
+            options,
+          );
+          createPreloadResource(preloadDocument, href, preloadProps);
+        }
+      }
+      return;
+    }
+
+    switch (as) {
+      case 'style': {
+        const precedence = options.precedence || 'default';
+        let resource = styleResources.get(href);
+        if (resource) {
+          if (__DEV__) {
+            const latestProps = stylePropsFromPreinitOptions(
+              href,
+              precedence,
+              options,
+            );
+            validateStyleResourceDifference(resource.props, latestProps);
+          }
+        } else {
+          const resourceProps = stylePropsFromPreinitOptions(
+            href,
+            precedence,
+            options,
+          );
+          resource = createStyleResource(
+            currentDocument,
+            href,
+            precedence,
+            resourceProps,
+          );
+        }
+        acquireResource(resource);
```

Comment:
I wonder if we actually want to immediately acquire these or if we want to first preload them and then inject them at the next time some other styles are updated so that they get batched.

Reply:
Interesting. So on next commit phase? in the same place we acquire today (layout but moving to before snapshop soon)?

Reply:
The same time we'd remove any styles if we did. Like when there's recalc anyway.

Maybe we don't do this yet since we're changing these phases anyway. But the reason I thought of it is because we need to inject either a preload or the style here so that we've marked this after adding it to the resource set. The code structure right now makes this a little tricky to guarantee.

Basically every querySelector needs to be followed by a synchronous insertion if it's missing, so that another run from Fizz, another root or something can find it.